ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/server/akka/StreamingServiceLifecycleManagement.scala

```diff
@@ -58,4 +64,39 @@ trait StreamingServiceLifecycleManagement extends AutoCloseable {
       }
     }
   }
+
+  implicit class ScalaPbOptimizationsFlow[ScalaPbMessage <: GeneratedMessage with AnyRef, Mat](
+      val original: Source[ScalaPbMessage, Mat]
+  ) {
+
+    /** Optimization for gRPC stream throughput.
+      *
+      * gRPC internal logic marshalls the protobuf response payloads sequentially before
+      * sending them over the wire (see io.grpc.ServerCallImpl.sendMessageInternal), imposing as limit
+      * the maximum marshalling throughput of a payload type.
+      *
+      * We've observed empirically that ScalaPB-generated message classes have associated marshallers
+      * with significant latencies when encoding complex payloads (e.g. [[com.daml.ledger.api.v1.transaction_service.GetTransactionTreesResponse]]),
+      * with the gRPC marshalling bottleneck appearing in some performance tests.
+      *
+      * As an alleviation of the problem, we can leverage the fact that ScalaPB message classes have the serializedSize value memoized,
+      * (see [[scalapb.GeneratedMessage.writeTo]]), whose computation is roughly half of the entire marshalling step.
+      *
+      * This optimization method takes advantage of the memoized value and forces the message's serializedSize computation,
+      * roughly doubling the maximum theoretical ScalaPB stream throughput over the gRPC server layer.
+      *
+      * @return A new source with precomputed serializedSize for the [[scalapb.GeneratedMessage]]
+      */
+    def precomputeSerializedSize: Source[ScalaPbMessage, Mat] =
+      original.map { msg =>
+        // At ScalaPB 0.11.8, computation of serializedSize is thread-safe but the memoization mechanism
+        // which this optimization is relying on is not.
+        // Hence, synchronize on the message to ensure across-threads visibility of the memoized value,
+        // with minimal performance impact since the monitor is un-contended.
+        msg.synchronized {
+          val _ = msg.serializedSize
+        }
```

Comment:
Maybe I misunderstood something, but if you synchronise on the message itself - the other call should be also synchronised on the same message in order to safely access memoized value. That means scalapb code which is serializing this message into byte array and as part of this serialization calls `serializedSize` - it has to synchronise on the message itself as well, otherwise this `synchronized` does not have any effect apart from wiping out CPU caches before and after entering this closure.

Reply:
Not sure, but maybe I'm not getting your point right. If you:
* are talking about two threads racing while calling this method, this should not happen. The calls should be strictly sequential (first this one here) and, then in the Sink Akka stage, the thread actually calling the marshalling (which includes the call to `serializedSize` again).
* are talking about the memory visibility of the changes performed here, this is exactly what I am aiming with the synchronized. `synchronized` here is meant to act as a memory barrier for the fields touched by the current thread and force their flush to main memory (as opposed to keeping them around in this thread's cache memory section). The next thread (the marshalling thread) would then see the updated values in main memory.

Reply:
@tudor-da I am not sure @skisel-da is in the wrong in here. For what I "know" from old is either you use volatile, or synchronize on both write and read. Also based on some quick search'n'read here: https://stackoverflow.com/questions/39929963/how-synchronized-and-volatile-work-in-java-memory-model, it seem to me the synchronize does not wipe caches, but rather it makes sure in the semaphore you are working from memory, which would reinforce this concern above.

On the sunny side AFAICT the worst thing what can happen here is that it is (maybe partially) will be recomputed again ;) :thinking:

Reply:
Thank you, you have actually clarified my original concern. 

Still I am not sure `synchronized` serves the purpose of your intention.  The **reader** thread will see the changes regardless of this `synchronized` block as this is guaranteed by akka, being in different stages and making sure they're and memory barriers are set correctly. Or am I wrong here?

I see currently `synchronized`  will help if you push the same object multiple times to this stream and memoization must be serialized to make sure object is not corrupted.

Reply:
@nmarton-da Yes, @skisel-da is right that there is still a slim chance that the **reader** thread somehow sees the non-memoized value (since it was written before, at its creation at least). I mentioned this in the last paragraph in my comment above.

If we want to be super defensive, we can actually memoize these objects on creation, which is a couple of places, but we would be mixing this grpc optimization with business logic. 

> Still I am not sure synchronized serves the purpose of your intention. The reader thread will see the changes regardless of this synchronized block as this is guaranteed by akka, being in different stages and making sure they're and memory barriers are set correctly. Or am I wrong here?

No, it is not guaranteed that the **reader** will see the changes unless that variable is also in a synchronized block (not aware of any guarantees provided by Akka on memory visibility of objects). I though that's what you (correctly) suggested initially.

Reply:
I would be also worried a little bit about accidental locking in here. AFAICS currently all messages will be different objects for fan-out streams because API serialization happens after IMFO. But this puts in some hidden locking trap as soon we would want to fan out identical streams.
Did you try without the synchronize? Is it much worse that way? If not I would consider not synchronizing due to these two concerns above :thinking:

Reply:
I've tried without synchronization and afair the results were the same. But then we are at the mercy of the cache coherency mechanism of the CPUs.

Reply:
> I've tried without synchronization and afair the results were the same. But then we are at the mercy of the cache coherency mechanism of the CPUs.

It seems to me we anyway are, since we are not synchronizing on reads.

Reply:
My understanding is that akka is backed by actors, so all these stages of stream will be executed in actor. By default all these simple stages like `map` would fuse these stages together and remove async boundaries between them. So it is kinda thread-safe to execute mutable code as it will be anyway done sequentially. 

Here we are introducing async boundary which means it will be implemented by another actor backed by queue and proper synchronisation. This means that mutable changes done in a previous stage should be published to a next stage.

Reply:
Thanks for the link, it actually helps to resolve the concern. So the key is actually in the quote you shared (basically to think twice before using synchronized) as well as here:

> if a message is sent to an actor (e.g. by another actor). In most cases messages are immutable, but if that message is not a properly constructed immutable object, without a “happens before” rule, it would be possible for the receiver to see partially initialized data structures and possibly even values out of thin air (longs/doubles).

It looks to me we need to trigger "happens before" rule on the **read thread**. I thought this has to be synchronized on the same object as well in order to make these changes visible...

Reply:
yeah, synchronized is dangerous if you misuse it. But I wasn't planning to :trollface: we can synchronize on the read as well, but it's uglier.

Reply:
Here are the two options for ensuring visibility on the **reader**:
* Synchronize on the message in the **reader**: 13639c029f23e39b6b3a0144108127766afe7687
* PR with precomputing the serializedSize for the messages on creation: https://github.com/digital-asset/daml/pull/15408

Reply:
Well, it looks much better as it does not have any concurrency primitive used, at least I did not notice. Does it actually optimize a throughput as well?